MetadataExtractor/IO/RandomAccessStream.cs

```diff
@@ -0,0 +1,578 @@
+        {
+            count = (int)ValidateIndex(index, count, isSequential, allowPartial);
+
+            // This bypasses a lot of checks particularly when the input was a byte[]
+            // TODO: good spot to try Span<T>
+            if (p_isStreamFinished && p_chunks.Count == 1)
+            {
+                Array.Copy(p_chunks[0], (int)index, buffer, 0, count);
+                return count;
+            }
+
+            var remaining = count;                      // how many bytes are requested
+            var fromOffset = (int)index;
+            var toIndex = offset > 0 ? offset : 0;
+            while (remaining != 0)
+            {
+                var fromChunkIndex = fromOffset / p_chunkLength;     // chunk integer key
+                var fromInnerIndex = fromOffset % p_chunkLength;     // index inside the chunk to start reading
+                var length = Math.Min(remaining, p_chunkLength - fromInnerIndex);
+                var chunk = p_chunks[fromChunkIndex];
+                Array.Copy(chunk, fromInnerIndex, buffer, toIndex, length);
+                remaining -= length;
+                fromOffset += length;
+                toIndex += length;
+            }
+
+            return toIndex - offset;
+        }
+
+        /// <summary>Returns an unsigned byte at an index in the sequence.</summary>
+        /// <returns>the 8 bit int value, between 0 and 255</returns>
+        /// <param name="index">position within the data buffer to read byte</param>
+        /// <param name="isSequential">flag indicating if caller is using sequential access</param>
+        /// <exception cref="BufferBoundsException"/>
+        /// <exception cref="IOException"/>
+        public byte GetByte(long index, bool isSequential)
+        {
+            return GetByte(index, isSequential, true);
+        }
+
+        /// <summary>Returns an unsigned byte at an index in the sequence.</summary>
+        /// <returns>the 8 bit int value, between 0 and 255</returns>
+        /// <param name="index">position within the data buffer to read byte</param>
+        /// <param name="isSequential">flag indicating if caller is using sequential access</param>
+        /// <param name="validateIndex">allows for skipping validation if already done by the caller</param>
+        /// <exception cref="BufferBoundsException"/>
+        /// <exception cref="IOException"/>
+        private byte GetByte(long index, bool isSequential, bool validateIndex)
+        {
+            if(validateIndex)
+                ValidateIndex(index, 1, isSequential);
+
+            // This bypasses a lot of checks particularly when the input was a byte[]
+            if (p_isStreamFinished && p_chunks.Count == 1)
+                return p_chunks[0][index];
+
+            var chunkIndex = index / p_chunkLength;
+            var innerIndex = index % p_chunkLength;
+
+            if (p_chunks.ContainsKey(chunkIndex))
+                return p_chunks[chunkIndex][innerIndex];
```

Comment:
I am a little concerned that these operations, when repeated for every byte consumed, will degrade performance. Rather than speculate, and while we have both the new/old implementations on the same branch, can we add some benchmarks that compare before/after?

I would recommend using BenchmarkDotNet for this. It's a really amazing library. I can help set that up if you'd like.

For scenarios, I'd look at sequential non-indexed access, sequential indexed access, random access.

Reply:
This loop is an almost verbatim copy of similar methods in IndexedCapturingReader. The only difference really is they get called in all cases instead of being implicitly requested by the caller (through using that reader class).

I can try to setup some benchmarks if you want. I'll probably need some guidance on BenchmarkDotNet if that's recommended.